llvm/include/llvm/Support/Casting.h

```diff
@@ -614,12 +610,11 @@ template <typename T> struct ValueIsPresent<std::optional<T>> {
   static inline decltype(auto) unwrapValue(std::optional<T> &t) { return *t; }
 };
 
-// If something is "nullable" then we just compare it to nullptr to see if it
-// exists.
+// Specialization for types convertible to bool.
 template <typename T>
-struct ValueIsPresent<T, std::enable_if_t<IsNullable<T>>> {
+struct ValueIsPresent<T, std::enable_if_t<std::is_constructible_v<bool, T>>> {
```

Comment:
Should we keep the check that T can be constructed with `nullptr`? There are types that can be cast to bool but are not pointer-like.

Reply:
I thought about it, but figured that checking whether the type is convertible to bool exactly answers the question "does an object have a value". This may not be desired for non-class types (e.g. `int`) but those shouldn't really be used in this context.

Reply:
But this enable_if guards both `isPresent` and `unwrapValue` -- don't we need extra care for the latter?

Reply:
Anyway, I'm going to approve this as-is, use your judgement here

Reply:
> But this enable_if guards both `isPresent` and `unwrapValue` -- don't we need extra care for the latter?

Maybe, although constructibility from nullptr_t doesn't guarantee that the default implementation of unwrapValue would fit.

As I noted in the other comment, I think it would be best if we provided explicit specializations for each type like we do with `simplify_type`, for example. Possibly provide a generic one for `T *`, too. I would't commit myself to implementing this though, I've already shown my expertise in metaprogramming.